test/sam.c

```diff
@@ -2149,6 +2149,27 @@ static void test_bam_set1_write_and_read_back()
     ks_free(&ks);
 }
 
+static void test_cigar_api(void)
+{
+    uint32_t *buf = NULL;
+    char *cig = "*";
+    char *end;
+    uint32_t m = 0;
+    int32_t n;
+    n = sam_parse_cigar(cig, &end, &buf, &m);
+    VERIFY(n == 0 && m == 0 && (end-cig) == 1, "failed to parse undefined CIGAR");
+    cig = "2M3X1I10M5D";
+    n = sam_parse_cigar(cig, &end, &buf, &m);
+    VERIFY(n == 5 && m > 0 && (end-cig) == 11, "failed to parse CIGAR string: 2M3X1I10M5D");
+    n = sam_parse_cigar("722M15D187217376188323783284M67I", NULL, &buf, &m);
```

Comment:
I'm not sure what the point is of this test.  187217376188323783284 is 68 bits long. 

Cigars are held inside htslib as uint32_t with 4 bits being the type, so the maximum length of cigar op we currently handle is 268435455 (2^28-1).   CRAM could handle longer lengths (if #976 got merged), but I think we rejected that idea as the in-memory htslib struct would need fixing first.

I'm sure we discussed changing the cigar type when we were doing all of our sweeping ABI changes and it was rejected, so I think we're committed to having excessively long cigar lengths as a series of concatenated smaller lengths using the same op type.  (I don't see a reason to support that though until we get reads long enough and accurate enough to generate such data.  We're considerably far off that still.)

Reply:
This does raise the question though why the test isn't:

```
VERIFY(n == -1, "failed to flag CIGAR string with long op length: 722M15D187217376188323783284M67I");
```

The documentation for `sam_parse_cigar` says -1 on error and "number of processed CIGAR operators" otherwise.  If the CIGAR has been rejected (even if syntacically valid) then it should probably return -1 in this case.